File: standard/erp_solution_ethiopia_standards/report/human_resource_informations/human_resource_informations.py


# import frappe
# from frappe import _
# from frappe.utils import getdate, nowdate


# def execute(filters=None):
# 	filters = filters or {}
# 	columns = get_columns()
# 	data = get_data(filters)
# 	return columns, data


# def get_columns():
# 	return [
# 		{"label": _("Employee"), "fieldname": "name", "fieldtype": "Link",
# 			"options": "Employee", "width": 110},
# 		{"label": _("Employee Number"), "fieldname": "employee_number", "fieldtype": "Data", "width": 110},
# 		{"label": _("Full Name"), "fieldname": "employee_name", "fieldtype": "Data", "width": 170},
# 		{"label": _("Gender"), "fieldname": "gender", "fieldtype": "Link",
# 			"options": "Gender", "width": 90},
# 		{"label": _("Date of Birth"), "fieldname": "date_of_birth", "fieldtype": "Date", "width": 100},
# 		{"label": _("Age"), "fieldname": "age", "fieldtype": "Int", "width": 60},
# 		{"label": _("Birth Date (EC)"), "fieldname": "custom_birth_date_ec", "fieldtype": "Data", "width": 100},
# 		{"label": _("Date of Joining"), "fieldname": "date_of_joining", "fieldtype": "Date", "width": 105},
# 		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link",
# 			"options": "Company", "width": 150},
# 		{"label": _("Branch"), "fieldname": "custom_branch", "fieldtype": "Link",
# 			"options": "Company Branch", "width": 120},
# 		{"label": _("Department"), "fieldname": "department", "fieldtype": "Link",
# 			"options": "Department", "width": 160},
# 		{"label": _("Designation"), "fieldname": "designation", "fieldtype": "Link",
# 			"options": "Designation", "width": 140},
# 		{"label": _("Employment Type"), "fieldname": "employment_type", "fieldtype": "Link",
# 			"options": "Employment Type", "width": 120},
# 		{"label": _("Project"), "fieldname": "custom_project", "fieldtype": "Link",
# 			"options": "Project", "width": 140},
# 		{"label": _("Grade"), "fieldname": "grade", "fieldtype": "Link",
# 			"options": "Employee Grade", "width": 100},
# 		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 90},
# 		{"label": _("Mobile"), "fieldname": "cell_number", "fieldtype": "Data", "width": 120},
# 		{"label": _("User ID"), "fieldname": "user_id", "fieldtype": "Link",
# 			"options": "User", "width": 160},
# 		{"label": _("Basic Salary"), "fieldname": "custom_basic_salary", "fieldtype": "Currency", "width": 110},
# 		{"label": _("Pension"), "fieldname": "custom_pension", "fieldtype": "Check", "width": 80},
# 		{"label": _("TIN Number"), "fieldname": "custom_tin_number", "fieldtype": "Data", "width": 110},
# 		{"label": _("Pension Number"), "fieldname": "custom_pension_number", "fieldtype": "Data", "width": 110},
# 		{"label": _("Relieving Date"), "fieldname": "relieving_date", "fieldtype": "Date", "width": 100},
# 	]


# def get_data(filters):
# 	conditions, values, date_fieldname = get_conditions(filters)

# 	fields = [
# 		"name", "employee_number", "employee_name", "gender", "date_of_birth",
# 		"custom_birth_date_ec", "date_of_joining", "company", "custom_branch",
# 		"department", "designation", "employment_type", "custom_project", "grade",
# 		"status", "cell_number", "user_id", "custom_basic_salary", "custom_pension",
# 		"custom_tin_number", "custom_pension_number", "relieving_date",
# 	]

# 	query = """
# 		SELECT {fields}
# 		FROM `tabEmployee`
# 		WHERE 1=1 {conditions}
# 		ORDER BY {date_field} ASC, employee_name ASC
# 	""".format(fields=", ".join(fields), conditions=conditions, date_field=date_fieldname)

# 	data = frappe.db.sql(query, values, as_dict=1)

# 	for row in data:
# 		row["age"] = calculate_age(row.get("date_of_birth"))

# 	return data


# def calculate_age(dob):
# 	"""Return completed years of age as of today, based on date_of_birth."""
# 	if not dob:
# 		return None

# 	dob = getdate(dob)
# 	today = getdate(nowdate())

# 	age = today.year - dob.year
# 	if (today.month, today.day) < (dob.month, dob.day):
# 		age -= 1

# 	return age


# def get_conditions(filters):
# 	"""Build parameterized SQL conditions (avoids SQL injection risk)."""
# 	conditions = ""
# 	values = {}

# 	# Employee Status: All Employee / Active Employee / Left Employee
# 	# Decides both the status filter (if any) and which date column
# 	# (date_of_joining vs relieving_date) the From/To Date range applies to.
# 	employee_status = filters.get("employee_status") or "All Employee"

# 	if employee_status == "Active Employee":
# 		conditions += " AND status = %(emp_status)s"
# 		values["emp_status"] = "Active"
# 		date_fieldname = "date_of_joining"
# 	elif employee_status == "Left Employee":
# 		conditions += " AND status = %(emp_status)s"
# 		values["emp_status"] = "Left"
# 		date_fieldname = "relieving_date"
# 	else:  # All Employee
# 		date_fieldname = "date_of_joining"

# 	# Date range filter, applied against the date column chosen above
# 	# (Gregorian dates arrive already converted from EC by the client)
# 	if filters.get("from_date"):
# 		conditions += " AND {0} >= %(from_date)s".format(date_fieldname)
# 		values["from_date"] = filters.get("from_date")

# 	if filters.get("to_date"):
# 		conditions += " AND {0} <= %(to_date)s".format(date_fieldname)
# 		values["to_date"] = filters.get("to_date")

# 	if filters.get("company"):
# 		conditions += " AND company = %(company)s"
# 		values["company"] = filters.get("company")

# 	if filters.get("branch"):
# 		conditions += " AND custom_branch = %(branch)s"
# 		values["branch"] = filters.get("branch")

# 	if filters.get("department"):
# 		conditions += " AND department = %(department)s"
# 		values["department"] = filters.get("department")

# 	if filters.get("designation"):
# 		conditions += " AND designation = %(designation)s"
# 		values["designation"] = filters.get("designation")

# 	if filters.get("employment_type"):
# 		conditions += " AND employment_type = %(employment_type)s"
# 		values["employment_type"] = filters.get("employment_type")

# 	if filters.get("project"):
# 		conditions += " AND custom_project = %(project)s"
# 		values["project"] = filters.get("project")

# 	if filters.get("grade"):
# 		conditions += " AND grade = %(grade)s"
# 		values["grade"] = filters.get("grade")

# 	# Legacy detailed status filter (Active/Inactive/Suspended/Left) only
# 	# applies when "All Employee" is selected, to avoid conflicting with
# 	# the status condition already set above.
# 	if filters.get("status") and employee_status == "All Employee":
# 		conditions += " AND status = %(status)s"
# 		values["status"] = filters.get("status")

# 	return conditions, values, date_fieldname
import frappe
from frappe import _
from frappe.utils import getdate, nowdate


def execute(filters=None):
	filters = filters or {}
	columns = get_columns()
	data = get_data(filters)
	report_summary = get_report_summary(data)
	chart = get_chart_data(data)
	return columns, data, None, chart, report_summary


def get_columns():
	return [
		{"label": _("Employee"), "fieldname": "name", "fieldtype": "Link",
			"options": "Employee", "width": 110},
		{"label": _("Employee Number"), "fieldname": "employee_number", "fieldtype": "Data", "width": 110},
		{"label": _("Full Name"), "fieldname": "employee_name", "fieldtype": "Data", "width": 170},
		{"label": _("Gender"), "fieldname": "gender", "fieldtype": "Link",
			"options": "Gender", "width": 90},
		{"label": _("Date of Birth"), "fieldname": "date_of_birth", "fieldtype": "Date", "width": 100},
		{"label": _("Age"), "fieldname": "age", "fieldtype": "Int", "width": 60},
		{"label": _("Birth Date (EC)"), "fieldname": "custom_birth_date_ec", "fieldtype": "Data", "width": 100},
		{"label": _("Date of Joining"), "fieldname": "date_of_joining", "fieldtype": "Date", "width": 105},
		{"label": _("Company"), "fieldname": "company", "fieldtype": "Link",
			"options": "Company", "width": 150},
		{"label": _("Branch"), "fieldname": "custom_branch", "fieldtype": "Link",
			"options": "Company Branch", "width": 120},
		{"label": _("Department"), "fieldname": "department", "fieldtype": "Link",
			"options": "Department", "width": 160},
		{"label": _("Designation"), "fieldname": "designation", "fieldtype": "Link",
			"options": "Designation", "width": 140},
		{"label": _("Employment Type"), "fieldname": "employment_type", "fieldtype": "Link",
			"options": "Employment Type", "width": 120},
		{"label": _("Project"), "fieldname": "custom_project", "fieldtype": "Link",
			"options": "Project", "width": 140},
		{"label": _("Grade"), "fieldname": "grade", "fieldtype": "Link",
			"options": "Employee Grade", "width": 100},
		{"label": _("Status"), "fieldname": "status", "fieldtype": "Data", "width": 90},
		{"label": _("Mobile"), "fieldname": "cell_number", "fieldtype": "Data", "width": 120},
		{"label": _("User ID"), "fieldname": "user_id", "fieldtype": "Link",
			"options": "User", "width": 160},
		{"label": _("Basic Salary"), "fieldname": "custom_basic_salary", "fieldtype": "Currency", "width": 110},
		{"label": _("Pension"), "fieldname": "custom_pension", "fieldtype": "Check", "width": 80},
		{"label": _("TIN Number"), "fieldname": "custom_tin_number", "fieldtype": "Data", "width": 110},
		{"label": _("Pension Number"), "fieldname": "custom_pension_number", "fieldtype": "Data", "width": 110},
		{"label": _("Relieving Date"), "fieldname": "relieving_date", "fieldtype": "Date", "width": 100},
	]


def get_data(filters):
	conditions, values, date_fieldname = get_conditions(filters)

	fields = [
		"name", "employee_number", "employee_name", "gender", "date_of_birth",
		"custom_birth_date_ec", "date_of_joining", "company", "custom_branch",
		"department", "designation", "employment_type", "custom_project", "grade",
		"status", "cell_number", "user_id", "custom_basic_salary", "custom_pension",
		"custom_tin_number", "custom_pension_number", "relieving_date",
	]

	query = """
		SELECT {fields}
		FROM `tabEmployee`
		WHERE 1=1 {conditions}
		ORDER BY {date_field} ASC, employee_name ASC
	""".format(fields=", ".join(fields), conditions=conditions, date_field=date_fieldname)

	data = frappe.db.sql(query, values, as_dict=1)

	for row in data:
		row["age"] = calculate_age(row.get("date_of_birth"))

	return data


def calculate_age(dob):
	"""Return completed years of age as of today, based on date_of_birth."""
	if not dob:
		return None

	dob = getdate(dob)
	today = getdate(nowdate())

	age = today.year - dob.year
	if (today.month, today.day) < (dob.month, dob.day):
		age -= 1

	return age


def get_report_summary(data):
	"""Cards shown at the top of the report (Total / Male / Female / Other)."""
	total = len(data)
	male = len([d for d in data if d.get("gender") == "ወንድ"])
	female = len([d for d in data if d.get("gender") == "ሴት"])
	other = total - male - female

	summary = [
		{"value": total, "label": _("Total Employees"), "datatype": "Int", "indicator": "Blue"},
		{"value": male, "label": _("Male"), "datatype": "Int", "indicator": "Blue"},
		{"value": female, "label": _("Female"), "datatype": "Int", "indicator": "Pink"},
	]

	if other:
		summary.append({
			"value": other,
			"label": _("Other / Not Specified"),
			"datatype": "Int",
			"indicator": "Grey",
		})

	return summary


def get_chart_data(data):
	"""Donut chart of gender distribution."""
	gender_counts = {}
	for row in data:
		g = row.get("gender") or _("Not Specified")
		gender_counts[g] = gender_counts.get(g, 0) + 1

	return {
		"data": {
			"labels": list(gender_counts.keys()),
			"datasets": [{"name": _("Employees"), "values": list(gender_counts.values())}],
		},
		"type": "donut",
		"colors": ["#5e64ff", "#ff5858", "#ffa00a", "#98d85b"],
	}


def get_conditions(filters):
	"""Build parameterized SQL conditions (avoids SQL injection risk)."""
	conditions = ""
	values = {}

	# Employee Status: All Employee / Active Employee / Left Employee
	# Decides both the status filter (if any) and which date column
	# (date_of_joining vs relieving_date) the From/To Date range applies to.
	employee_status = filters.get("employee_status") or "All Employee"

	if employee_status == "Active Employee":
		conditions += " AND status = %(emp_status)s"
		values["emp_status"] = "Active"
		date_fieldname = "date_of_joining"
	elif employee_status == "Left Employee":
		conditions += " AND status = %(emp_status)s"
		values["emp_status"] = "Left"
		date_fieldname = "relieving_date"
	else:  # All Employee
		date_fieldname = "date_of_joining"

	# Date range filter, applied against the date column chosen above
	# (Gregorian dates arrive already converted from EC by the client)
	if filters.get("from_date"):
		conditions += " AND {0} >= %(from_date)s".format(date_fieldname)
		values["from_date"] = filters.get("from_date")

	if filters.get("to_date"):
		conditions += " AND {0} <= %(to_date)s".format(date_fieldname)
		values["to_date"] = filters.get("to_date")

	if filters.get("company"):
		conditions += " AND company = %(company)s"
		values["company"] = filters.get("company")

	if filters.get("branch"):
		conditions += " AND custom_branch = %(branch)s"
		values["branch"] = filters.get("branch")

	if filters.get("department"):
		conditions += " AND department = %(department)s"
		values["department"] = filters.get("department")

	if filters.get("designation"):
		conditions += " AND designation = %(designation)s"
		values["designation"] = filters.get("designation")

	if filters.get("employment_type"):
		conditions += " AND employment_type = %(employment_type)s"
		values["employment_type"] = filters.get("employment_type")

	if filters.get("project"):
		conditions += " AND custom_project = %(project)s"
		values["project"] = filters.get("project")

	if filters.get("grade"):
		conditions += " AND grade = %(grade)s"
		values["grade"] = filters.get("grade")

	if filters.get("gender"):
		conditions += " AND gender = %(gender)s"
		values["gender"] = filters.get("gender")

	# Legacy detailed status filter (Active/Inactive/Suspended/Left) only
	# applies when "All Employee" is selected, to avoid conflicting with
	# the status condition already set above.
	if filters.get("status") and employee_status == "All Employee":
		conditions += " AND status = %(status)s"
		values["status"] = filters.get("status")

	return conditions, values, date_fieldname

File: standard/erp_solution_ethiopia_standards/report/human_resource_informations/human_resource_informations.js
// Copyright (c) 2026, ERP Solution Ethiopia PLC and contributors
// For license information, please see license.txt

/* ------------------------------------------------------------------ *
 *  Ethiopian <-> Gregorian calendar conversion (Julian Day Number)
 *  Standard proleptic conversion, Amete Mihret epoch.
 *  NOTE: This block must stay at the top of the file — the filter
 *  defaults below call toEC() at load time, before the object literal
 *  is even built, so these must already be initialized by then.
 * ------------------------------------------------------------------ */

const JD_EPOCH_OFFSET_AMETE_MIHRET = 1723856;

function ethiopicToJDN(year, month, day) {
	return day + (month - 1) * 30 + (year - 1) * 365 + Math.floor(year / 4) + JD_EPOCH_OFFSET_AMETE_MIHRET - 1;
}

function jdnToEthiopic(jdn) {
	const r = (jdn - JD_EPOCH_OFFSET_AMETE_MIHRET) % 1461;
	const n = (r % 365) + 365 * Math.floor(r / 1460);
	const year =
		4 * Math.floor((jdn - JD_EPOCH_OFFSET_AMETE_MIHRET) / 1461) + Math.floor(r / 365) - Math.floor(r / 1460);
	const month = Math.floor(n / 30) + 1;
	const day = (n % 30) + 1;
	return [year, month, day];
}

function gregorianToJDN(year, month, day) {
	const a = Math.floor((month - 14) / 12);
	return (
		Math.floor((1461 * (year + 4800 + a)) / 4) +
		Math.floor((367 * (month - 2 - 12 * a)) / 12) -
		Math.floor((3 * Math.floor((year + 4900 + a) / 100)) / 4) +
		day -
		32075
	);
}

function jdnToGregorian(jdn) {
	const a = jdn + 32044;
	const b = Math.floor((4 * a + 3) / 146097);
	const c = a - Math.floor((146097 * b) / 4);
	const d = Math.floor((4 * c + 3) / 1461);
	const e = c - Math.floor((1461 * d) / 4);
	const m = Math.floor((5 * e + 2) / 153);
	const day = e - Math.floor((153 * m + 2) / 5) + 1;
	const month = m + 3 - 12 * Math.floor(m / 10);
	const year = 100 * b + d - 4800 + Math.floor(m / 10);
	return [year, month, day];
}

function toEC(gcArray) {
	const [year, month, day] = gcArray;
	return jdnToEthiopic(gregorianToJDN(year, month, day));
}

function toGC(ecArray) {
	const [year, month, day] = ecArray;
	return jdnToGregorian(ethiopicToJDN(year, month, day));
}

/* ------------------------------------------------------------------ *
 *  Report definition
 * ------------------------------------------------------------------ */

let syncingDates = false; // guards against the GC<->EC round-trip feedback loop

// Compute EC defaults up front so the EC filters are populated the moment
// the filter panel renders — no dependency on onload timing or filter order.
const defaultToDateGC = frappe.datetime.get_today();
const defaultFromDateGC = frappe.datetime.add_months(defaultToDateGC, -12); // last year -> today

function gcStringToEcString(gcDateStr) {
	const [y, m, d] = toEC(gcDateStr.split("-").map(Number));
	return `${d.toString().padStart(2, "0")}/${m.toString().padStart(2, "0")}/${y}`;
}

frappe.query_reports["Human Resource Informations"] = {
	"filters": [
		{
			"fieldname": "employee_status",
			"label": __("Employee Status"),
			"fieldtype": "Select",
			"options": ["All Employee", "Active Employee", "Left Employee"],
			"default": "All Employee",
			"reqd": 1,
			"change": function () {
				updateDateFilterLabels(frappe.query_report);
			},
		},
		{
			"fieldname": "ec_from_date_ec",
			"label": __("From Date (E.C.)"),
			"fieldtype": "Data",
			"reqd": 1,
			"default": gcStringToEcString(defaultFromDateGC),
			"change": function () { ConvertFromDate(frappe.query_report); },
		},
		{
			"fieldname": "ec_to_date_ec",
			"label": __("To Date (E.C.)"),
			"fieldtype": "Data",
			"reqd": 1,
			"default": gcStringToEcString(defaultToDateGC),
			"change": function () { ConvertToDate(frappe.query_report); },
		},
		{
			"fieldname": "from_date",
			"label": __("From Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1,
			"default": defaultFromDateGC,
		},
	
		{
			"fieldname": "to_date",
			"label": __("To Date"),
			"fieldtype": "Date",
			"width": "80",
			"reqd": 1,
			"default": defaultToDateGC,
		},
	
		{
			"fieldname": "company",
			"label": __("Company"),
			"fieldtype": "Link",
			"options": "Company",
			"default": frappe.defaults.get_user_default("Company"),
		},
		{
			"fieldname": "branch",
			"label": __("Company Branch"),
			"fieldtype": "Link",
			"options": "Company Branch",
		},
		{
			"fieldname": "department",
			"label": __("Department"),
			"fieldtype": "Link",
			"options": "Department",
			"get_query": function () {
				var company = frappe.query_report.get_filter_value("company");
				return { filters: { company: company } };
			},
		},
		{
			"fieldname": "designation",
			"label": __("Designation"),
			"fieldtype": "Link",
			"options": "Designation",
		},
		{
			"fieldname": "employment_type",
			"label": __("Employment Type"),
			"fieldtype": "Link",
			"options": "Employment Type",
		},
		{
			"fieldname": "project",
			"label": __("Project"),
			"fieldtype": "Link",
			"options": "Project",
		},
		{
			"fieldname": "grade",
			"label": __("Grade"),
			"fieldtype": "Link",
			"options": "Employee Grade",
		},
		{
			"fieldname": "gender",
			"label": __("Gender"),
			"fieldtype": "Link",
			"options": "Gender",
		},
		{
			"fieldname": "status",
			"label": __("Status"),
			"fieldtype": "Select",
			"options": ["", "Active", "Inactive", "Suspended", "Left"],
		},
	],

	formatter: function (value, row, column, data, default_formatter) {
		value = default_formatter(value, row, column, data);
		if (column.fieldname === "age" && data && data.age !== undefined && data.age !== null) {
			value = `<span style="color: green; font-weight: 600;">${data.age}</span>`;
		}
		return value;
	},

	onload: function (report) {
		// Defaults already cover the initial EC values now. This just keeps
		// GC and EC in sync if something else (e.g. a saved filter set,
		// or a future change) updates from_date/to_date without going
		// through ConvertFromDate/ConvertToDate first.
		syncFromGCtoEC(report);
		updateDateFilterLabels(report);

		// Forces a clean run now that all reqd filters (including the EC
		// ones) are guaranteed to have values.
		report.refresh();
	},
};

/* Converts current from_date/to_date (GC) into the EC filters, guarded
 * so it doesn't trigger ConvertFromDate/ConvertToDate and bounce back. */
function syncFromGCtoEC(report) {
	syncingDates = true;

	const fromDate = report.get_filter_value("from_date");
	const toDate = report.get_filter_value("to_date");

	if (fromDate) {
		report.set_filter_value("ec_from_date_ec", gcStringToEcString(fromDate));
	}
	if (toDate) {
		report.set_filter_value("ec_to_date_ec", gcStringToEcString(toDate));
	}

	syncingDates = false;
}

function ConvertFromDate(report) {
	if (syncingDates) return; // skip when triggered programmatically, not by the user

	const ecDate = report.get_filter_value("ec_from_date_ec");
	if (ecDate) {
		const [day, month, year] = ecDate.split("/").map(Number);
		const [y, m, d] = toGC([year, month, day]);
		report.set_filter_value("from_date", `${y}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`);
	}
}

function ConvertToDate(report) {
	if (syncingDates) return;

	const ecDate = report.get_filter_value("ec_to_date_ec");
	if (ecDate) {
		const [day, month, year] = ecDate.split("/").map(Number);
		const [y, m, d] = toGC([year, month, day]);
		report.set_filter_value("to_date", `${y}-${String(m).padStart(2, "0")}-${String(d).padStart(2, "0")}`);
	}
}

/* Relabels the date filters so users know whether the range applies to
 * Joining Date or Relieving Date, based on the Employee Status filter. */
function updateDateFilterLabels(report) {
	const employeeStatus = report.get_filter_value("employee_status");
	const isLeft = employeeStatus === "Left Employee";

	const labels = {
		from_date: isLeft ? __("From Date (Relieving)") : __("From Date (Joining)"),
		to_date: isLeft ? __("To Date (Relieving)") : __("To Date (Joining)"),
		ec_from_date_ec: isLeft ? __("From Date (Relieving) (E.C.)") : __("From Date (Joining) (E.C.)"),
		ec_to_date_ec: isLeft ? __("To Date (Relieving) (E.C.)") : __("To Date (Joining) (E.C.)"),
	};

	Object.keys(labels).forEach(function (fieldname) {
		const filter = report.get_filter(fieldname);
		if (filter) {
			filter.df.label = labels[fieldname];
			filter.refresh();
		}
	});
}